File: ruby/easy/pascals_triangle.rb
# @param {Integer} num_rows
# @return {Integer[][]}
def generate(num_rows)
    res = []
    generate_triangle(0, num_rows, res)
    return res
end

def generate_triangle(i, num_rows, res)
    return if(i == num_rows)
    temp = [1]
    if(i > 0)
        prev = res[i-1]
        for j in 0..prev.length-1
            temp.push((j == prev.length-1) ? 1 : prev[j]+prev[j+1])
        end
    end
    res.push(temp)
    generate_triangle(i+1, num_rows, res)
end

File: javascript/easy/pascalsTriangle.js
/**
 * @param {number} numRows
 * @return {number[][]}
 */
var generate = function(numRows) {
    let queue = [], i = 0, res = [];
    if(numRows == 0) return res;
    queue.unshift([1]);
    while(i < numRows && queue.length > 0) {
        let level = queue.pop(), temp = [];
        res.push(level);
        for(let j = 0; j < level.length-1; j++) {
            temp.push(level[j] + level[j+1]);
        }
        temp.unshift(1); temp.push(1);
        queue.unshift(temp);
        i++;
    }
    return res;
};
